File: WWF/Tutorials/Tutorial2/StateMachineWorkflow/Exercise2/Task5/VB/StateMachineWorkflow/StateMachineWorkflow.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------
Imports System
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Collections
Imports System.Drawing
Imports System.Workflow.ComponentModel.Compiler
Imports System.Workflow.ComponentModel.Serialization
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design
Imports System.Workflow.Runtime
Imports System.Workflow.Activities
Imports System.Workflow.Activities.Rules

Namespace Microsoft.Samples.Workflow.Tutorials.StateMachineWorkflow
    <Serializable()> _
       Public Class NewOrderEventArgs : Inherits ExternalDataEventArgs
        Private orderItemId As Guid
        Private orderItem As String
        Private orderQuantity As Integer

        Public Sub New(ByVal itemId As Guid, ByVal item As String, _
            ByVal quantity As Integer)
            MyBase.New(itemId)
            Me.orderItemId = itemId
            Me.orderItem = item
            Me.orderQuantity = quantity
        End Sub

        Public Property ItemId() As Guid
            Get
                Return orderItemId
            End Get
            Set(ByVal value As Guid)
                orderItemId = Value
            End Set
        End Property

        Public Property Item() As String
            Get
                Return orderItem
            End Get
            Set(ByVal value As String)
                orderItem = Value
            End Set
        End Property

        Public Property Quantity() As Integer
            Get
                Return orderQuantity
            End Get
            Set(ByVal value As Integer)
                orderQuantity = Value
            End Set
        End Property
    End Class

    <ExternalDataExchange()> _
       Public Interface IOrderingService
        Sub ItemStatusUpdate(ByVal orderId As Guid, ByVal newStatus As String)
        Event NewOrder As EventHandler(Of NewOrderEventArgs)
    End Interface

    Public Class OrderProcessingWorkflow : Inherits StateMachineWorkflowActivity
        Private WaitingForOrderStateActivity As StateActivity
        Private OrderProcessingStateActivity As StateActivity
        Private OrderCompletedStateActivity As StateActivity

        Private eventDriven1 As EventDrivenActivity
        Private newOrderExternalEvent As HandleExternalEventActivity
        Private updatestatusOrderReceived As CallExternalMethodActivity
        Private setStateActivityOrderProcessing As SetStateActivity
        Public receivedOrderDetails As NewOrderEventArgs

        ' OrderProcessingStateActivity Activities
        Private initializeOrderOpenStateActivity As StateInitializationActivity
        Private invokeProcessingNewOrderStatusUpdate As CallExternalMethodActivity
        Private invokeOrderProcessedStatusUpdate As CallExternalMethodActivity
        Private setStateActivityOrderCompleted As SetStateActivity

        Public orderId As Guid
        Private orderItem As String
        Private orderQuantity As Integer
        Public orderItemStatus As String

        Public ReadOnly Property Id() As Guid
            Get
                Return Me.orderId
            End Get
        End Property

        Public ReadOnly Property ItemStatus() As String
            Get
                Return Me.orderItemStatus
            End Get
        End Property

        Public ReadOnly Property Item() As String
            Get
                Return Me.orderItem
            End Get
        End Property

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub InitializeComponent()
            Me.CanModifyActivities = True
            Dim activitybind1 As System.Workflow.ComponentModel.ActivityBind = _
                New System.Workflow.ComponentModel.ActivityBind()
            Dim activitybind2 As System.Workflow.ComponentModel.ActivityBind = _
                New System.Workflow.ComponentModel.ActivityBind()
            Dim activitybind3 As System.Workflow.ComponentModel.ActivityBind = _
                New System.Workflow.ComponentModel.ActivityBind()
            Dim workflowparameterbinding1 As _
                System.Workflow.ComponentModel.WorkflowParameterBinding = _
                New System.Workflow.ComponentModel.WorkflowParameterBinding()
            Dim workflowparameterbinding2 As _
                System.Workflow.ComponentModel.WorkflowParameterBinding = _
                New System.Workflow.ComponentModel.WorkflowParameterBinding()
            Dim workflowparameterbinding3 As _
            System.Workflow.ComponentModel.WorkflowParameterBinding = _
            New System.Workflow.ComponentModel.WorkflowParameterBinding()

            Me.WaitingForOrderStateActivity = New _
                System.Workflow.Activities.StateActivity()
            Me.OrderProcessingStateActivity = New _
                System.Workflow.Activities.StateActivity()
            Me.OrderCompletedStateActivity = New _
                System.Workflow.Activities.StateActivity()

            ' WaitingForOrderStateActivity Activities
            Me.eventDriven1 = New System.Workflow.Activities.EventDrivenActivity()
            Me.newOrderExternalEvent = New _
                System.Workflow.Activities.HandleExternalEventActivity()
            Me.updatestatusOrderReceived = New _
                System.Workflow.Activities.CallExternalMethodActivity()
            Me.setStateActivityOrderProcessing = New _
                System.Workflow.Activities.SetStateActivity()

            ' OrderProcessingStateActivity Activities
            Me.initializeOrderOpenStateActivity = New _
                System.Workflow.Activities.StateInitializationActivity()
            Me.invokeProcessingNewOrderStatusUpdate = New _
                System.Workflow.Activities.CallExternalMethodActivity()
            Me.invokeOrderProcessedStatusUpdate = New _
                System.Workflow.Activities.CallExternalMethodActivity()
            Me.setStateActivityOrderCompleted = New _
                System.Workflow.Activities.SetStateActivity()
            ' 
            ' WaitingForOrderStateActivity
            ' 
            Me.WaitingForOrderStateActivity.Activities.Add(Me.eventDriven1)
            Me.WaitingForOrderStateActivity.Name = "WaitingForOrderStateActivity"
            ' 
            ' OrderProcessingStateActivity
            ' 
            Me.OrderProcessingStateActivity.Activities.Add _
                (Me.initializeOrderOpenStateActivity)
            Me.OrderProcessingStateActivity.Name = "OrderProcessingStateActivity"
            ' 
            ' OrderCompletedStateActivity
            ' 
            Me.OrderCompletedStateActivity.Name = "OrderCompletedStateActivity"
            ' 
            ' eventDriven1
            ' 
            Me.eventDriven1.Activities.Add(Me.newOrderExternalEvent)
            Me.eventDriven1.Activities.Add(Me.updatestatusOrderReceived)
            Me.eventDriven1.Activities.Add(Me.setStateActivityOrderProcessing)
            Me.eventDriven1.Name = "eventDriven1"
            ' 
            ' newOrderExternalEvent
            ' 
            Me.newOrderExternalEvent.EventName = "NewOrder"
            Me.newOrderExternalEvent.InterfaceType = _
                GetType(StateMachineWorkflow.IOrderingService)
            Me.newOrderExternalEvent.Name = "newOrderExternalEvent"
            activitybind1.Name = "OrderProcessingWorkflow"
            activitybind1.Path = "receivedOrderDetails"
            workflowparameterbinding1.ParameterName = "e"
            workflowparameterbinding1.SetBinding( _
                System.Workflow.ComponentModel.WorkflowParameterBinding.ValueProperty, _
                (CType(activitybind1, System.Workflow.ComponentModel.ActivityBind)))
            Me.newOrderExternalEvent.ParameterBindings.Add(workflowparameterbinding1)
            ' 
            ' updatestatusOrderReceived
            ' 
            Me.updatestatusOrderReceived.InterfaceType = _
                GetType(StateMachineWorkflow.IOrderingService)
            Me.updatestatusOrderReceived.MethodName = "ItemStatusUpdate"
            Me.updatestatusOrderReceived.Name = "updatestatusOrderReceived"
            activitybind2.Name = "OrderProcessingWorkflow"
            activitybind2.Path = "orderId"
            workflowparameterbinding2.ParameterName = "orderId"
            workflowparameterbinding2.SetBinding( _
                System.Workflow.ComponentModel.WorkflowParameterBinding.ValueProperty, _
                (CType(activitybind2, System.Workflow.ComponentModel.ActivityBind)))
            activitybind3.Name = "OrderProcessingWorkflow"
            activitybind3.Path = "orderItemStatus"
            workflowparameterbinding3.ParameterName = "newStatus"
            workflowparameterbinding3.SetBinding( _
                System.Workflow.ComponentModel.WorkflowParameterBinding.ValueProperty, _
                (CType(activitybind3, System.Workflow.ComponentModel.ActivityBind)))
            Me.updatestatusOrderReceived.ParameterBindings.Add(workflowparameterbinding2)
            Me.updatestatusOrderReceived.ParameterBindings.Add(workflowparameterbinding3)
            AddHandler updatestatusOrderReceived.MethodInvoking, _
                AddressOf Me.OrderReceived
            ' 
            ' setStateActivityOrderProcessing
            ' 
            Me.setStateActivityOrderProcessing.Name = "setStateActivityOrderProcessing"
            Me.setStateActivityOrderProcessing.TargetStateName = _
                "OrderProcessingStateActivity"
            ' 
            ' initializeOrderOpenStateActivity
            ' 
            Me.initializeOrderOpenStateActivity.Activities.Add _
                (Me.invokeProcessingNewOrderStatusUpdate)
            Me.initializeOrderOpenStateActivity.Activities.Add _
                (Me.invokeOrderProcessedStatusUpdate)
            Me.initializeOrderOpenStateActivity.Activities.Add _
                (Me.setStateActivityOrderCompleted)
            Me.initializeOrderOpenStateActivity.Name = "initializeOrderOpenStateActivity"
            ' 
            ' invokeProcessingNewOrderStatusUpdate
            ' 
            Me.invokeProcessingNewOrderStatusUpdate.InterfaceType = _
                GetType(StateMachineWorkflow.IOrderingService)
            Me.invokeProcessingNewOrderStatusUpdate.MethodName = "ItemStatusUpdate"
            Me.invokeProcessingNewOrderStatusUpdate.ParameterBindings.Add _
                (workflowparameterbinding2)
            Me.invokeProcessingNewOrderStatusUpdate.ParameterBindings.Add _
                (workflowparameterbinding3)
            AddHandler invokeProcessingNewOrderStatusUpdate.MethodInvoking, _
                AddressOf ProcessNewOrder
            Me.invokeProcessingNewOrderStatusUpdate.Name = _
                "invokeProcessingNewOrderStatusUpdate"
            ' 
            ' invokeOrderProcessedStatusUpdate
            ' 
            Me.invokeOrderProcessedStatusUpdate.InterfaceType = _
                GetType(StateMachineWorkflow.IOrderingService)
            Me.invokeOrderProcessedStatusUpdate.MethodName = "ItemStatusUpdate"
            Me.invokeOrderProcessedStatusUpdate.ParameterBindings.Add _
                (workflowparameterbinding2)
            Me.invokeOrderProcessedStatusUpdate.ParameterBindings.Add _
                (workflowparameterbinding3)
            AddHandler invokeOrderProcessedStatusUpdate.MethodInvoking, _
                AddressOf FinalizeOrder
            Me.invokeOrderProcessedStatusUpdate.Name = "invokeOrderProcessedStatusUpdate"
            ' 
            ' setStateActivityOrderCompleted
            ' 
            Me.setStateActivityOrderCompleted.Name = "setStateActivityOrderCompleted"
            Me.setStateActivityOrderCompleted.TargetStateName = _
                "OrderCompletedStateActivity"
            ' 
            ' OrderProcessingWorkflow
            ' 
            Me.Activities.Add(Me.WaitingForOrderStateActivity)
            Me.Activities.Add(Me.OrderProcessingStateActivity)
            Me.Activities.Add(Me.OrderCompletedStateActivity)
            Me.InitialStateName = "WaitingForOrderStateActivity"
            Me.CompletedStateName = "OrderCompletedStateActivity"
            Me.DynamicUpdateCondition = Nothing
            Me.Name = "OrderProcessingWorkflow"
            Me.CanModifyActivities = False
        End Sub

        Private Sub OrderReceived(ByVal sender As Object, ByVal e As EventArgs)
            Me.orderId = receivedOrderDetails.ItemId
            Me.orderItem = receivedOrderDetails.Item
            Me.orderQuantity = receivedOrderDetails.Quantity
            Me.orderItemStatus = "Received order"
        End Sub

        Private Sub ProcessNewOrder(ByVal sender As Object, ByVal e As EventArgs)
            Me.orderItemStatus = "Processing order"
        End Sub

        Private Sub FinalizeOrder(ByVal sender As Object, ByVal e As EventArgs)
            Me.orderItemStatus = "Order processed"
        End Sub

    End Class
End Namespace

File: WWF/Tutorials/Tutorial2/StateMachineWorkflow/Exercise2/Task5/CS/StateMachineWorkflow/StateMachineWorkflow.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Collections;
using System.Drawing;
using System.Workflow.ComponentModel.Compiler;
using System.Workflow.ComponentModel.Serialization;
using System.Workflow.ComponentModel;
using System.Workflow.ComponentModel.Design;
using System.Workflow.Runtime;
using System.Workflow.Activities;
using System.Workflow.Activities.Rules;

namespace Microsoft.Samples.Workflow.Tutorials.StateMachineWorkflow
{
    [Serializable]
    public class NewOrderEventArgs : ExternalDataEventArgs
    {
        private Guid orderItemId;
        private string orderItem;
        private int orderQuantity;

        public NewOrderEventArgs(Guid itemId, string item, int quantity)
            : base(itemId)
        {
            this.orderItemId = itemId;
            this.orderItem = item;
            this.orderQuantity = quantity;
        }

        public Guid ItemId
        {
            get { return orderItemId; }
            set { orderItemId = value; }
        }

        public string Item
        {
            get { return orderItem; }
            set { orderItem = value; }
        }

        public int Quantity
        {
            get { return orderQuantity; }
            set { orderQuantity = value; }
        }
    }

    [ExternalDataExchange]
    public interface IOrderingService
    {
        void ItemStatusUpdate(Guid orderId, string newStatus);
        event EventHandler<NewOrderEventArgs> NewOrder;
    }

    public class OrderProcessingWorkflow : StateMachineWorkflowActivity
    {
        private StateActivity WaitingForOrderStateActivity;
        private StateActivity OrderProcessingStateActivity;
        private StateActivity OrderCompletedStateActivity;

        // WaitingForOrderStateActivity Activities
        private EventDrivenActivity eventDriven1;
        private HandleExternalEventActivity newOrderExternalEvent;
        private CallExternalMethodActivity updatestatusOrderReceived;
        private SetStateActivity setStateActivityOrderProcessing;
        public NewOrderEventArgs receivedOrderDetails = null;

        // OrderProcessingStateActivity Activities
        private StateInitializationActivity initializeOrderOpenStateActivity;
        private CallExternalMethodActivity invokeProcessingNewOrderStatusUpdate;
        private CallExternalMethodActivity invokeOrderProcessedStatusUpdate;
        private SetStateActivity setStateActivityOrderCompleted;

        public Guid orderId;
        private string orderItem;
        private int orderQuantity;
        public string orderItemStatus;

        public Guid Id
        {
            get
            {
                return this.orderId;
            }
        }

        public string ItemStatus
        {
            get
            {
                return this.orderItemStatus;
            }
        }
        
        public string Item
        {
            get
            {
                return this.orderItem;
            }
        }
        
        public OrderProcessingWorkflow()
        {
            InitializeComponent();
        }

        private void InitializeComponent()
        {
            this.CanModifyActivities = true;
            ActivityBind activitybind1 = new ActivityBind();
            ActivityBind activitybind2 = new ActivityBind();
            ActivityBind activitybind3 = new ActivityBind();
            WorkflowParameterBinding workflowparameterbinding1 = 
                new WorkflowParameterBinding();
            WorkflowParameterBinding workflowparameterbinding2 = 
                new WorkflowParameterBinding();
            WorkflowParameterBinding workflowparameterbinding3 = 
                new WorkflowParameterBinding();
            this.WaitingForOrderStateActivity = new StateActivity();
            this.OrderProcessingStateActivity = new StateActivity();
            this.OrderCompletedStateActivity = new StateActivity();

            // WaitingForOrderStateActivity child activities
            this.eventDriven1 = new EventDrivenActivity();
            this.newOrderExternalEvent = new HandleExternalEventActivity();
            this.updatestatusOrderReceived = new CallExternalMethodActivity();
            this.setStateActivityOrderProcessing = new SetStateActivity();

            // OrderProcessingStateActivity Activities
            this.initializeOrderOpenStateActivity = new StateInitializationActivity();
            this.invokeProcessingNewOrderStatusUpdate = new CallExternalMethodActivity();
            this.invokeOrderProcessedStatusUpdate = new CallExternalMethodActivity();
            this.setStateActivityOrderCompleted = new SetStateActivity();
            // 
            // WaitingForOrderStateActivity
            // 
            this.WaitingForOrderStateActivity.Activities.Add(this.eventDriven1);
            this.WaitingForOrderStateActivity.Name = "WaitingForOrderStateActivity";
            // 
            // OrderProcessingStateActivity
            // 
            this.OrderProcessingStateActivity.Activities.Add
                (this.initializeOrderOpenStateActivity);
            this.OrderProcessingStateActivity.Name = "OrderProcessingStateActivity";
            // 
            // OrderCompletedStateActivity
            // 
            this.OrderCompletedStateActivity.Name = "OrderCompletedStateActivity";
            // 
            // eventDriven1
            // 
            this.eventDriven1.Activities.Add(this.newOrderExternalEvent);
            this.eventDriven1.Activities.Add(this.updatestatusOrderReceived);
            this.eventDriven1.Activities.Add(this.setStateActivityOrderProcessing);
            this.eventDriven1.Name = "eventDriven1";
            // 
            // newOrderExternalEvent
            // 
            this.newOrderExternalEvent.EventName = "NewOrder";
            this.newOrderExternalEvent.InterfaceType = typeof(IOrderingService);
            activitybind1.Name = "OrderProcessingWorkflow";
            activitybind1.Path = "receivedOrderDetails";
            workflowparameterbinding1.ParameterName = "e";
            workflowparameterbinding1.SetBinding(
                System.Workflow.ComponentModel.WorkflowParameterBinding.ValueProperty, 
                ((System.Workflow.ComponentModel.ActivityBind)(activitybind1)));
            this.newOrderExternalEvent.ParameterBindings.Add(workflowparameterbinding1);
            this.newOrderExternalEvent.Name = "newOrderExternalEvent";
            // 
            // updatestatusOrderReceived
            // 
            this.updatestatusOrderReceived.InterfaceType = typeof(IOrderingService);
            this.updatestatusOrderReceived.MethodName = "ItemStatusUpdate";
            this.updatestatusOrderReceived.Name = "updatestatusOrderReceived";
            activitybind2.Name = "OrderProcessingWorkflow";
            activitybind2.Path = "orderId";
            workflowparameterbinding2.ParameterName = "orderId";
            workflowparameterbinding2.SetBinding(WorkflowParameterBinding.ValueProperty,
                ((System.Workflow.ComponentModel.ActivityBind)(activitybind2)));
            activitybind3.Name = "OrderProcessingWorkflow";
            activitybind3.Path = "orderItemStatus";
            workflowparameterbinding3.ParameterName = "newStatus";
            workflowparameterbinding3.SetBinding(WorkflowParameterBinding.ValueProperty,
                ((System.Workflow.ComponentModel.ActivityBind)(activitybind3)));
            this.updatestatusOrderReceived.ParameterBindings.Add
                (workflowparameterbinding2);
            this.updatestatusOrderReceived.ParameterBindings.Add
                (workflowparameterbinding3);
            this.updatestatusOrderReceived.MethodInvoking += 
                new System.EventHandler(this.OrderReceived);
            // 
            // setStateActivityOrderProcessing
            // 
            this.setStateActivityOrderProcessing.Name = 
                "setStateActivityOrderProcessing";
            this.setStateActivityOrderProcessing.TargetStateName = 
                "OrderProcessingStateActivity";
            // 
            // initializeOrderOpenStateActivity
            // 
            this.initializeOrderOpenStateActivity.Activities.Add
                (this.invokeProcessingNewOrderStatusUpdate);
            this.initializeOrderOpenStateActivity.Activities.Add
                (this.invokeOrderProcessedStatusUpdate);
            this.initializeOrderOpenStateActivity.Activities.Add
                (this.setStateActivityOrderCompleted);
            this.initializeOrderOpenStateActivity.Name = 
                "initializeOrderOpenStateActivity";
            // 
            // invokeProcessingNewOrderStatusUpdate
            // 
            this.invokeProcessingNewOrderStatusUpdate.InterfaceType = 
                typeof(IOrderingService);
            this.invokeProcessingNewOrderStatusUpdate.MethodName = "ItemStatusUpdate";
            this.invokeProcessingNewOrderStatusUpdate.ParameterBindings.Add
                (workflowparameterbinding2);
            this.invokeProcessingNewOrderStatusUpdate.ParameterBindings.Add
                (workflowparameterbinding3);
            this.invokeProcessingNewOrderStatusUpdate.MethodInvoking += 
                new System.EventHandler(this.ProcessNewOrder);
            this.invokeProcessingNewOrderStatusUpdate.Name = 
                "invokeProcessingNewOrderStatusUpdate";
            // 
            // invokeOrderProcessedStatusUpdate
            // 
            this.invokeOrderProcessedStatusUpdate.InterfaceType = 
                typeof(IOrderingService);
            this.invokeOrderProcessedStatusUpdate.MethodName = "ItemStatusUpdate";
            this.invokeOrderProcessedStatusUpdate.ParameterBindings.Add
                (workflowparameterbinding2);
            this.invokeOrderProcessedStatusUpdate.ParameterBindings.Add
                (workflowparameterbinding3);
            this.invokeOrderProcessedStatusUpdate.MethodInvoking += 
                new System.EventHandler(this.FinalizeOrder);
            this.invokeOrderProcessedStatusUpdate.Name = 
                "invokeOrderProcessedStatusUpdate";
            // 
            // setStateActivityOrderCompleted
            // 
            this.setStateActivityOrderCompleted.Name = 
                "setStateActivityOrderCompleted";
            this.setStateActivityOrderCompleted.TargetStateName = 
                "OrderCompletedStateActivity";
            // 
            // OrderProcessingWorkflow
            // 
            this.Activities.Add(this.WaitingForOrderStateActivity);
            this.Activities.Add(this.OrderProcessingStateActivity);
            this.Activities.Add(this.OrderCompletedStateActivity);
            this.InitialStateName = "WaitingForOrderStateActivity";
            this.CompletedStateName = "OrderCompletedStateActivity";
            this.DynamicUpdateCondition = null;
            this.Name = "OrderProcessingWorkflow";
            this.CanModifyActivities = false;
        }

        private void OrderReceived(object sender, EventArgs e)
        {
            this.orderId = receivedOrderDetails.ItemId;
            this.orderItem = receivedOrderDetails.Item;
            this.orderQuantity = receivedOrderDetails.Quantity;
            this.orderItemStatus = "Received order";
        }

        private void ProcessNewOrder(object sender, EventArgs e)
        {
            this.orderItemStatus = "Processing order";
        }

        void FinalizeOrder(object sender, EventArgs e)
        {
            this.orderItemStatus = "Order processed";
        }
    }
}
